### PR TITLE
8260106: Shenandoah: refactor reference updating closures and related code

### PR DESCRIPTION
**Pull request by shipilev**
We have a block in `ShenandoahHeap::maybe_update_with_forwarded` that is irrelevant after JDK-8231086. Additionally, "resolve and update" paths are really only used by STW GCs, and thus do not require atomic updates. This leads to considerable simplifications in the code, and improves performance on the common paths (especially in fastdebug builds that drop many irrelevant asserts).

Additional testing:
 - [x] `hotspot_gc_shenandoah`
 - [x] `tier1` with Shenandoah
 - [x] `tier2` with Shenandoah

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed

### Issue
 * [JDK-8260106](https://bugs.openjdk.java.net/browse/JDK-8260106): Shenandoah: refactor reference updating closures and related code

### Reviewers
 * [Zhengyu Gu](https://openjdk.java.net/census#zgu) (@zhengyu123 - **Reviewer**) ⚠️ Review applies to 4f30251de7736bc48865bba949996be3640375e9
 * [Roman Kennke](https://openjdk.java.net/census#rkennke) (@rkennke - **Reviewer**) ⚠️ Review applies to 4f30251de7736bc48865bba949996be3640375e9

### Download
`$ git fetch https://git.openjdk.java.net/jdk pull/2166/head:pull/2166`
`$ git checkout pull/2166`

### CODE REVIEW

**Review by zhengyu123 — APPROVED** [review]
Looks good. 

Please update copyright years of  shenandoahHeap.inline.hpp and shenandoahOopClosures.hpp

File: src/hotspot/share/gc/shenandoah/shenandoahHeap.inline.hpp [hunk 1]
```diff
@@ -100,45 +100,33 @@ inline oop ShenandoahHeap::update_with_forwarded_not_null(T* p, oop obj) {
...
-      }
+    if (in_collection_set(obj)) {
+      // Boundary condition: when evacuation fails, there are objects in collection
+      // set that are not really forwarded. We can still go and try CAS-update them
+      // (uselessly) to simplify the common path.
+      shenandoah_assert_forwarded_except(p, obj, cancelled_gc());
+      oop fwd = ShenandoahBarrierSet::resolve_forwarded_not_null(obj);
+      shenandoah_assert_not_in_cset_except(p, fwd, cancelled_gc());
+
+      // Sanity check: we are should not be updating the cset regions themselves,
```
Comment (rkennke) [inline src/hotspot/share/gc/shenandoah/shenandoahHeap.inline.hpp:120]: Typo: excess 'are' (I think)

**Review by rkennke — APPROVED** [review]
Nice! Few nits below.

File: src/hotspot/share/gc/shenandoah/shenandoahHeap.inline.hpp (hunk 1 above)
Comment (shipilev) [reply src/hotspot/share/gc/shenandoah/shenandoahHeap.inline.hpp]: Resolved, along with a few other touchups.

**Review by rkennke — APPROVED** [review]
Looks good! Thanks!

**Comment by shipilev** [comment]
I updated the patch a bit after `RESOLVE` mode removal. I think `NO_UPDATE` and `STW_UPDATE` reads better. This also highlights that STW closures are really doing non-CAS updates.

**Review by rkennke — APPROVED** [review]
Looks good to me! Thanks!

**Comment by shipilev** [comment]
This kinda keeps creeping: I just realized we need to sweep in the "normal" update references in the same thing. I'll make this a draft and keep working on it.

File: src/hotspot/share/gc/shenandoah/shenandoahOopClosures.hpp [hunk 2]
```diff
@@ -188,41 +188,37 @@ class ShenandoahMarkRefsMetadataDedupClosure : public ShenandoahMarkRefsSuperClo
 };
 
 class ShenandoahUpdateRefsSuperClosure : public BasicOopIterateClosure {
-private:
-  ShenandoahHeap* _heap;
-
 protected:
-  template <class T, UpdateRefsMode UPDATE_MODE>
-  inline void work(T *p);
+  ShenandoahHeap* _heap;
```
Comment (rkennke) [inline src/hotspot/share/gc/shenandoah/shenandoahOopClosures.hpp:192]: While moving it around, maybe make it Shenandoah* const too?

**Review by rkennke — APPROVED** [review]
This is indeed much better! I only have a very minor comment,

File: src/hotspot/share/gc/shenandoah/shenandoahOopClosures.hpp (hunk 2 above)
Comment (shipilev) [reply src/hotspot/share/gc/shenandoah/shenandoahOopClosures.hpp:192]: Right! Did in new changeset.

**Review by rkennke — APPROVED** [review]
Looks good! Thanks!

**Review by zhengyu123 — APPROVED** [review]
Still good.

**Pull request closed without merging**